### PR TITLE
Fixed point (bit-accurate with Loihi) and floating point ProcessModels for LIF process

### PR DESCRIPTION
**Pull request by srrisbud**
The fixed point ProcessModel, which is bit-accurate with Loihi, is leanest ProcessModel of such behaviour. Features like axonal delay, refractory period, etc. are not supported.

### CODE REVIEW

File: tests/proc/lif/test_models.py [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+        """Receive spikes and store in an internal variable"""
+        spk_in = self.s_in.recv()
+        self.spk_data[self.current_ts-1, :] = spk_in
+
+
+class TestLIFProcessModels(unittest.TestCase):
+    """Tests for all ProcessModels of LIF"""
+    def test_float_pm_no_decay(self):
+        """
+        Set up bias = 1 * 2**1 = 2. and threshold = 4. Without any current
```
Comment (awintel) [inline tests/proc/lif/test_models.py:145]: Often also useful to document unit test code inline to show intent so it's easier to understand in half a year.

File: tests/proc/lif/test_models.py [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+        # Run 1 timestep at a time and collect state variable u
+        for j in range(num_steps):
+            lif.run(condition=rcnd, run_cfg=rcfg)
+            lif_v.append(lif.v.get()[0])
+        lif.stop()
+        expected_v_timeseries = [128., 192., 224., 240., 248., 252., 254., 255.]
+        print(expected_v_timeseries)
+        self.assertListEqual(expected_v_timeseries, lif_v)
+
+    def test_bitacc_pm_no_decay(self):
```
Comment (awintel) [inline tests/proc/lif/test_models.py:241]: This should go into a separate TestCase

File: tests/proc/lif/test_models.py [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+        lif.run(condition=rcnd, run_cfg=rcfg)
+        spk_data_through_run = spr.spk_data.get()
+        lif.stop()
+        expected_spk_data = np.zeros((num_steps, shape[0]))
+        expected_spk_data[1:10:2, :] = 1.
+        self.assertTrue(np.all(expected_spk_data == spk_data_through_run))
+
+    def test_float_pm_impulse_du(self):
+        """
+        Send a single spike of magnitude 128 with a current decay of 0.5 and
```
Comment (awintel) [inline tests/proc/lif/test_models.py:174]: Docstring should say what it checks and maybe why this needs to be checked. 
Inline comments could also be used to explain the how along the way.

File: tests/proc/lif/test_models.py [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+        lif.stop()
+        expected_u_timeseries = [1 << (13-j) for j in range(8)]
+        expected_float_u = [1 << (7-j) for j in range(8)]
+        self.assertListEqual(expected_u_timeseries, lif_u)
+        self.assertListEqual(expected_float_u, np.right_shift(np.array(
+            lif_u), 6).tolist())
+
+    def test_bitacc_pm_impulse_dv(self):
+        """
+        Send a single activation of magnitude 128 with a voltage decay of
```
Comment (awintel) [inline tests/proc/lif/test_models.py:310]: Hopefully now you see, that it would be more worthwhile to start with WHAT this unit test is about and WHY we are doing this rather than deep diving into the nitty gritty details of the math. Again this might be more useful after the WHAT and WHY or through inline comments.
This will go over most people's head when they read this for the first time.

File: tests/proc/lif/test_process.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+        """Tests instantiation of LIF"""
+        lif = LIF(shape=(100,),
+                  du=100*np.ones((100,), dtype=np.float),
+                  dv=np.ones((100,), dtype=np.float),
+                  bias=2*np.ones((100,), dtype=np.float),
+                  bias_exp=np.ones((100,), dtype=np.float),
+                  vth=np.ones((100,), dtype=np.float))
+
+        self.assertEqual(lif.shape, (100,))
+        self.assertListEqual(lif.du.init.tolist(), 100 * [100.])
```
Comment (awintel) [inline tests/proc/lif/test_process.py:21]: I guess we should update the get() method to return the init value when called before compilation.

File: tests/proc/lif/test_models.py [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+    s_in: PyInPort = LavaPyType(PyInPort.VEC_DENSE, bool, precision=1)
+    spk_data: np.ndarray = LavaPyType(np.ndarray, int, precision=1)
+
+    def run_spk(self):
+        """Receive spikes and store in an internal variable"""
+        spk_in = self.s_in.recv()
+        self.spk_data[self.current_ts-1, :] = spk_in
+
+
+class TestLIFProcessModels(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (awintel) [reply tests/proc/lif/test_models.py:141]: ... coming back here after going through the whole file...
If you don't have time before the release, please add unit tests to test a few more of the key features of the bit-accurate model like overflow handling, rounding towards zero.

**Review by awintel — CHANGES_REQUESTED** [review]
Looks good but a few things require changes if there's still time or filing of separate issues.

File: tests/proc/lif/test_models.py [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+        lif.run(condition=rcnd, run_cfg=rcfg)
+        spk_data_through_run = spr.spk_data.get()
+        lif.stop()
+        expected_spk_data = np.zeros((num_steps, shape[0]))
+        expected_spk_data[3:10:4, :] = 1
+        self.assertTrue(np.all(expected_spk_data == spk_data_through_run))
+
+    def test_bitacc_pm_impulse_du(self):
+        """
+        Send a single activation of magnitude 128 with a current decay of
```
Comment (drager-intel) [inline tests/proc/lif/test_models.py:271]: Start the docstring with "Test of Loihi 1 bit accurate LIF dynamics with external input and a decay of the integrated current. "

File: tests/proc/lif/test_models.py [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+        2047 and voltage decay of 0. Current state variable decays by half every
```
Comment (drager-intel) [inline tests/proc/lif/test_models.py:272]: Perhaps add "Current decay can take values 0 to 4095, so a current decay of 2047 is equivalent to decaying by half in a floating point implementation of the dynamics."

File: tests/proc/lif/test_models.py (hunk 4 above)
Comment (drager-intel) [inline tests/proc/lif/test_models.py:310]: Begin the docstring with "Test of Loihi 1 bit accurate LIF dynamics with external input and a decay of the integrated voltage."

File: tests/proc/lif/test_models.py [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+        expected_u_timeseries = [1 << (13-j) for j in range(8)]
+        expected_float_u = [1 << (7-j) for j in range(8)]
+        self.assertListEqual(expected_u_timeseries, lif_u)
+        self.assertListEqual(expected_float_u, np.right_shift(np.array(
+            lif_u), 6).tolist())
+
+    def test_bitacc_pm_impulse_dv(self):
+        """
+        Send a single activation of magnitude 128 with a voltage decay of
+        2048 and current decay of 0. Voltage state variable increases from
```
Comment (drager-intel) [inline tests/proc/lif/test_models.py:311]: Remind the user "Voltage decay can take values 0 to 4095, so a voltage decay of 2048 is equivalent to decaying by half in a floating point implementation of the dynamics." 
Also, we used 2047 for testing a current decay of 1/2. Realize 4095 isn't even divisible by 2 so we have to choose 2047 or 2048, but is there a reason the choice is inconsistent between the current and the voltage decay tests?

**Review by drager-intel — CHANGES_REQUESTED** [review]

File: lava/magma/core/decorator.py [hunk 10]
```diff
@@ -102,3 +102,65 @@ def decorate_process_model(cls: type):
...
+    args: keywords that tag a ProcessModel
+
+    Returns
+    -------
+    Decorated class
+
+    Examples
+    --------
+    @implements(proc=ExampleProcess)
+    @tag('bit-accurate', 'loihi')
```
[the remark this answers is not in the document]
Comment (srrisbud) [reply lava/magma/core/decorator.py:134]: This is from an older commit.. for some reason github made this a part of this PR

File: lava/magma/core/decorator.py [hunk 11]
```diff
@@ -102,3 +102,65 @@ def decorate_process_model(cls: type):
...
+    """
+
+    arg_list = list(args)
+    tags = []
+
+    for arg in arg_list:
+        if isinstance(arg, str):
+            tags.append(arg)
+        else:
+            raise AssertionError("Invalid input to the 'tags' decorator. "
```
[the remark this answers is not in the document]
Comment (srrisbud) [reply lava/magma/core/decorator.py:149]: This is from an older commit.. for some reason github made this a part of this PR

File: lava/magma/core/decorator.py [hunk 12]
```diff
@@ -102,3 +102,65 @@ def decorate_process_model(cls: type):
...
+                                 "Valid input should be comma-separated "
+                                 "keywords as strings")
```
[the remark this answers is not in the document]
Comment (srrisbud) [reply lava/magma/core/decorator.py:151]: This is from an older commit.. for some reason github made this a part of this PR

File: lava/proc/lif/models.py [hunk 13]
```diff
@@ -7,29 +7,116 @@
...
 from lava.proc.lif.process import LIF
 
 
 @implements(proc=LIF, protocol=LoihiProtocol)
 @requires(CPU)
-class PyLifModel(PyLoihiProcessModel):
+@tag('floating_pt')
+class PyLifModel1(PyLoihiProcessModel):
+    a_in: PyInPort = LavaPyType(PyInPort.VEC_DENSE, np.float)
+    s_out: PyOutPort = LavaPyType(PyOutPort.VEC_DENSE, bool, precision=1)
```
[the remark this answers is not in the document]
Comment (srrisbud) [reply lava/proc/lif/models.py]: It is. Included there just for clarity. Can remove it as well. Should not make any difference.

File: tests/proc/lif/test_process.py (hunk 5 above)
Comment (srrisbud) [reply tests/proc/lif/test_process.py]: So this is not relevant to this code then.

File: lava/magma/compiler/builder.py [hunk 14]
```diff
@@ -337,7 +337,7 @@ def build(self):
             if issubclass(lt.cls, np.ndarray):
                 var = lt.cls(v.shape, lt.d_type)
                 var[:] = v.value
-            elif issubclass(lt.cls, int):
+            elif issubclass(lt.cls, int) or issubclass(lt.cls, float):
```
[the remark this answers is not in the document]
Comment (srrisbud) [reply lava/magma/compiler/builder.py:362]: Done.

File: tests/proc/lif/test_models.py (hunk 1 above)
Comment (srrisbud) [reply tests/proc/lif/test_models.py]: Done.

File: tests/proc/lif/test_models.py (hunk 2 above)
Comment (srrisbud) [reply tests/proc/lif/test_models.py]: Done.

File: tests/proc/lif/test_models.py (hunk 3 above)
Comment (srrisbud) [reply tests/proc/lif/test_models.py]: Done.

File: tests/proc/lif/test_models.py (hunk 4 above)
Comment (srrisbud) [reply tests/proc/lif/test_models.py]: Done.

File: lava/proc/lif/models.py [hunk 15]
```diff
@@ -7,29 +7,116 @@
...
+        self.v[:] = self.v * (1 - self.dv) + self.u + bias
+        s_out = self.v >= self.vth
+        self.v[s_out] = 0  # Reset voltage to 0
+        self.s_out.send(s_out)
+
+
+@implements(proc=LIF, protocol=LoihiProtocol)
+@requires(CPU)
+@tag('bit_accurate_loihi', 'fixed_pt')
+class PyLifModel2(PyLoihiProcessModel):
```
[the remark this answers is not in the document]
Comment (srrisbud) [reply lava/proc/lif/models.py]: Done.

File: tests/proc/lif/test_models.py [hunk 16]
```diff
@@ -0,0 +1,350 @@
...
+        self.shape = shape
+        self.s_in = InPort(shape=(shape[1],))
+        self.spk_data = Var(shape=shape, init=0)  # This Var expands with time
+
+
+@implements(proc=VecSendProcess, protocol=LoihiProtocol)
+@requires(CPU)
+# need the following tag to discover the ProcessModel using LifRunConfig
+@tag('floating_pt')
+class PyVecSendModel1(PyLoihiProcessModel):
```
[the remark this answers is not in the document]
Comment (srrisbud) [reply tests/proc/lif/test_models.py]: Done.

File: tests/proc/lif/test_models.py [hunk 17]
```diff
@@ -0,0 +1,350 @@
...
+            self.s_out.send(self.vec_to_send)
+        else:
+            self.s_out.send(np.zeros_like(self.vec_to_send))
+
+
+@implements(proc=VecSendProcess, protocol=LoihiProtocol)
+@requires(CPU)
+# need the following tag to discover the ProcessModel using LifRunConfig
+@tag('fixed_pt')
+class PyVecSendModel2(PyLoihiProcessModel):
```
[the remark this answers is not in the document]
Comment (srrisbud) [reply tests/proc/lif/test_models.py]: Done.

File: tests/proc/lif/test_models.py [hunk 18]
```diff
@@ -0,0 +1,350 @@
...
+            self.s_out.send(self.vec_to_send)
+        else:
+            self.s_out.send(np.zeros_like(self.vec_to_send))
+
+
+@implements(proc=VecRecvProcess, protocol=LoihiProtocol)
+@requires(CPU)
+# need the following tag to discover the ProcessModel using LifRunConfig
+@tag('floating_pt')
+class PySpkRecvModel1(PyLoihiProcessModel):
```
[the remark this answers is not in the document]
Comment (srrisbud) [reply tests/proc/lif/test_models.py]: Done.

File: tests/proc/lif/test_models.py [hunk 19]
```diff
@@ -0,0 +1,350 @@
...
+        """Receive spikes and store in an internal variable"""
+        spk_in = self.s_in.recv()
+        self.spk_data[self.current_ts-1, :] = spk_in
+
+
+@implements(proc=VecRecvProcess, protocol=LoihiProtocol)
+@requires(CPU)
+# need the following tag to discover the ProcessModel using LifRunConfig
+@tag('fixed_pt')
+class PySpkRecvModel2(PyLoihiProcessModel):
```
[the remark this answers is not in the document]
Comment (srrisbud) [reply tests/proc/lif/test_models.py]: Done.

File: tests/proc/lif/test_models.py [hunk 20]
```diff
@@ -0,0 +1,350 @@
...
+        self.shape = shape
+        self.num_steps = num_steps
+        self.vec_to_send = Var(shape=shape, init=vec_to_send)
+        self.send_at_times = Var(shape=(num_steps,), init=send_at_times)
+        self.s_out = OutPort(shape=shape)
+
+
+class VecRecvProcess(AbstractProcess):
+    """
+    Sink process that receives arbitrary vectors
```
[the remark this answers is not in the document]
Comment (awintel) [reply tests/proc/lif/test_models.py]: I'm saying it because of the difference in class name and "Sink" in docstring.

File: tests/proc/lif/test_models.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (srrisbud) [reply tests/proc/lif/test_models.py]: Done.

File: tests/proc/lif/test_models.py (hunk 7 above)
Comment (srrisbud) [reply tests/proc/lif/test_models.py]: Done.

File: tests/proc/lif/test_models.py (hunk 8 above)
Comment (srrisbud) [reply tests/proc/lif/test_models.py]: Done.

File: tests/proc/lif/test_models.py (hunk 4 above)
Comment (srrisbud) [reply tests/proc/lif/test_models.py]: Done.

File: tests/proc/lif/test_models.py (hunk 9 above)
Comment (srrisbud) [reply tests/proc/lif/test_models.py]: Done.

File: lava/proc/lif/models.py [hunk 21]
```diff
@@ -7,29 +7,116 @@
...
+        neg_current_limit = -np.left_shift(1, 23, dtype=np.int32)
+        pos_current_limit = np.left_shift(1, 23, dtype=np.int32) - 1
+        clipped_curr = np.clip(decayed_curr, neg_current_limit,
+                               pos_current_limit)
+        self.u[:] = clipped_curr
+        # Update voltage
+        # --------------
+        decay_const_v = self.dv + self.dm_offset
+        neg_voltage_limit = -np.left_shift(1, 23, dtype=np.int32) + 1
+        pos_voltage_limit = np.left_shift(1, 23, dtype=np.int32) - 1
```
[the remark this answers is not in the document]
Comment (srrisbud) [reply lava/proc/lif/models.py]: Well, after ignoring the refractory clutter, it is just clipping.

File: lava/proc/lif/models.py [hunk 22]
```diff
@@ -7,29 +7,116 @@
...
+        # Decaying voltage similar to current. See the comment above to
+        # understand the need for each of the operations below.
+        decayed_volt = np.right_shift(np.int64(self.v) * (np.left_shift(1, 12)
```
[the remark this answers is not in the document]
Comment (srrisbud) [reply lava/proc/lif/models.py]: Done.

File: lava/proc/lif/models.py [hunk 23]
```diff
@@ -7,29 +7,116 @@
...
+                                                          - decay_const_v), 12,
+                                      dtype=np.int32)
+        bias = np.left_shift(self.bias, self.bias_exp, dtype=np.int32)
```
[the remark this answers is not in the document]
Comment (srrisbud) [reply lava/proc/lif/models.py]: Done.

File: lava/proc/lif/models.py [hunk 24]
```diff
@@ -7,29 +7,116 @@
...
+        updated_volt = decayed_volt + self.u + bias
+        self.v[:] = np.clip(updated_volt, neg_voltage_limit, pos_voltage_limit)
+
+        # Spike when exceeds threshold
+        # ----------------------------
+        # In Loihi, user specified threshold is just the mantissa, with a
+        # constant exponent of 6
+        vth = np.left_shift(self.vth, 6, dtype=np.int32)
```
[the remark this answers is not in the document]
Comment (srrisbud) [reply lava/proc/lif/models.py]: Done.

File: lava/proc/lif/process.py [hunk 25]
```diff
@@ -8,23 +8,40 @@
...
-    v[t] = v[t-1] * (1-dv) + u[t] + bias
-    s_out = v[t] > vth
-    v[t] = v[t] - s_out*vth
+    LIF dynamics abstracts to:
+    u[t] = u[t-1] * (1-du) + a_in         # neuron current
+    v[t] = v[t-1] * (1-dv) + u[t] + bias  # neuron voltage
+    s_out = v[t] > vth                    # spike if threshold is exceeded
+    v[t] = 0                              # reset at spike
+
+    Parameters
```
[the remark this answers is not in the document]
Comment (srrisbud) [reply lava/proc/lif/process.py]: Explained only the ones that are user settable.

File: lava/proc/lif/process.py [hunk 26]
```diff
@@ -8,23 +8,40 @@
...
+    ----------
+    du: Inverse of decay time-constant for current decay.
+    dv: Inverse of decay time-constant for voltage decay.
+    bias: Mantissa part of neuron bias.
```
[the remark this answers is not in the document]
Comment (srrisbud) [reply lava/proc/lif/process.py]: Did it the other way around: floating point model uses bias_mant * (2 ** bias_exp) if provided.

File: lava/proc/lif/process.py [hunk 27]
```diff
@@ -8,23 +8,40 @@
...
+        du = kwargs.pop("du", 0)
+        dv = kwargs.pop("dv", 0)
+        bias = kwargs.pop("bias", 0)
+        bias_exp = kwargs.pop("bias_exp", 0)
+        vth = kwargs.pop("vth", 10)
+
+        self.shape = shape
         self.a_in = InPort(shape=shape)
         self.s_out = OutPort(shape=shape)
         self.u = Var(shape=shape, init=0)
```
[the remark this answers is not in the document]
Comment (srrisbud) [reply lava/proc/lif/process.py]: Good point. Will file an issue and add this as an enhancement later.

File: tests/proc/lif/test_models.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (srrisbud) [reply tests/proc/lif/test_models.py]: Removed Sink.

File: tests/proc/lif/test_models.py [hunk 28]
```diff
@@ -0,0 +1,350 @@
...
+                             vec_to_send=np.zeros(shape, dtype=np.float),
+                             send_at_times=np.ones((num_steps,), dtype=np.bool))
+        lif = LIF(shape=shape,
+                  du=0.,
+                  dv=0.,
+                  bias=np.ones(shape, dtype=np.float),
+                  bias_exp=np.ones(shape, dtype=np.float),
+                  vth=4.)
+        spr = VecRecvProcess(shape=(num_steps, shape[0]))
+        sps.out_ports.s_out.connect(lif.in_ports.a_in)
```
[the remark this answers is not in the document]
Comment (srrisbud) [reply tests/proc/lif/test_models.py]: Done.

File: tests/proc/lif/test_models.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (srrisbud) [reply tests/proc/lif/test_models.py]: Done.

File: tests/proc/lif/test_models.py [hunk 29]
```diff
@@ -0,0 +1,350 @@
...
+        for j in range(num_steps):
+            lif.run(condition=rcnd, run_cfg=rcfg)
+            lif_u.append(lif.u.get()[0])
+        lif.stop()
+        expected_u_timeseries = [2.**(7-j) for j in range(8)]
+        self.assertListEqual(expected_u_timeseries, lif_u)
+
+    def test_float_pm_impulse_dv(self):
+        """
+        Send a single spike of magnitude 128 with a voltage decay of 0.5 and
```
[the remark this answers is not in the document]
Comment (srrisbud) [reply tests/proc/lif/test_models.py]: Done.

File: tests/proc/lif/test_models.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (srrisbud) [reply tests/proc/lif/test_models.py]: Will create an issue and add these tests.

File: lava/proc/lif/models.py [hunk 30]
```diff
@@ -7,29 +7,116 @@
...
+                                                          - decay_const_u), 12,
+                                      dtype=np.int32)
+        # Hardware left-shifts synpatic input for MSB alignment
+        a_in_data = np.left_shift(a_in_data, 6)
+        # Add synptic input to decayed current
+        decayed_curr += a_in_data
+        # Check if value of current is within bounds
+        neg_current_limit = -np.left_shift(1, 23, dtype=np.int32)
+        pos_current_limit = np.left_shift(1, 23, dtype=np.int32) - 1
+        clipped_curr = np.clip(decayed_curr, neg_current_limit,
```
[the remark this answers is not in the document]
Comment (srrisbud) [reply lava/proc/lif/models.py]: Added a ToDo to address this later..

File: lava/proc/lif/models.py [hunk 31]
```diff
@@ -7,29 +7,116 @@
...
-from lava.magma.core.decorator import implements, requires
+from lava.magma.core.decorator import implements, requires, tag
 from lava.magma.core.model.py.model import PyLoihiProcessModel
 from lava.proc.lif.process import LIF
 
 
 @implements(proc=LIF, protocol=LoihiProtocol)
 @requires(CPU)
-class PyLifModel(PyLoihiProcessModel):
+@tag('floating_pt')
```
[the remark this answers is not in the document]
Comment (awintel) [reply lava/proc/lif/models.py]: That's exactly the point: Restrictions as in "well defined boundaries" what you can do rather than anything. It's not that you just tag something. There always must be something else on the other side that deals with the tag.

File: lava/proc/lif/models.py [hunk 32]
```diff
@@ -7,29 +7,116 @@
...
+    vth: float = LavaPyType(float, np.float)
+
+    def run_spk(self):
+        a_in_data = self.a_in.recv()
+        self.u[:] = self.u * (1 - self.du)
+        self.u[:] += a_in_data
+        bias = self.bias * (2**self.bias_exp)
+        self.v[:] = self.v * (1 - self.dv) + self.u + bias
+        s_out = self.v >= self.vth
+        self.v[s_out] = 0  # Reset voltage to 0
```
[the remark this answers is not in the document]
Comment (awintel) [reply lava/proc/lif/models.py]: So you literally talk about Python meta classes when talking about meta parameters to control what params exist?

Could be done I guess. But not sure if this is necessary. Question is why can it not just be a regular parameter to the constructor?

File: lava/proc/lif/models.py [hunk 33]
```diff
@@ -7,29 +7,116 @@
...
+        - refractory period after spiking
+        - axonal delays
+
+    Precisions of state variables
+    -----------------------------
+    du: unsigned 12-bit integer (0 to 4095)
+    dv: unsigned 12-bit integer (0 to 4095)
+    bias: signed 13-bit integer (-4096 to 4095). Mantissa part of neuron bias.
+    bias_exp: unsigned 3-bit integer (0 to 7). Exponent part of neuron bias.
+    vth: unsigned 17-bit integer (0 to 131071)
```
[the remark this answers is not in the document]
Comment (awintel) [reply lava/proc/lif/models.py]: Yes, see related comment in Process. Why restrict u, v from being initialized by the user? 
"Internal states" does not sound right... all states if defined as a Var are publicaly accessible.

File: lava/proc/lif/process.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (awintel) [reply lava/proc/lif/process.py]: What do you mean. Every Var is user-setable. my_lif.u.set([1, 2, 3]) will work.
